File: regex2.js
function solve(s, p) {
  let j = 0
  let flag = false
  for (let i = 0; i < p.length; i++) {
    if (flag === true) {
      if (p[i] === '*') {
        if (i === p.length - 1) {
          return true
        }
        continue
      }
      if (p[i] === '?') {
        j++
        continue
      }
      while (s[j] !== p[i]) {
        if (!s[j]) {
          return false
        }
        j++
      }
      while (s[j] === s[j + 1]) {
        j++
      }

      flag = false
    }

    if (s[j] === p[i] || p[i] === '?') {
      j++
      continue
    }
    if (p[i] === '*') {
      if (i === p.length - 1) {
        return true
      }
      flag = true
      continue
    }

    return false
  }
  if (j !== s.length) {
    return false
  }
  return true
}
const s = 'abab'
const p = '*ab'

console.log(solve(s, p))

File: regex2.cpp
#include <iostream>

using namespace std;
int F[2002][2002];

int matching(string s, string p)
{
  s = " " + s;
  p = " " + p;
  for (int j = 1; j < p.size(); j++)
  {
    if (p[j] != '*')
      break;
    for (int i = 0; i < s.size(); i++)
      F[i][j] = 1;
  }

  F[0][0] = 1;
  for (int i = 1; i < s.size(); i++)
  {
    for (int j = 1; j < p.size(); j++)
    {
      if (p[j] == '*')
      {
        F[i][j] |= F[i][j - 1];
        F[i][j] |= F[i - 1][j];
        F[i][j] |= F[i - 1][j - 1];
      }
      else if (p[j] == '?' || p[j] == s[i])
        F[i][j] |= F[i - 1][j - 1];
    }
  }
  return F[s.size() - 1][p.size() - 1];
}
int regex()
{
  string s = "abab";
  string p = "*?*ab";
  int isMatching = matching(s, p);

  cout << isMatching << endl;

  return 0;
}